Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm

```diff
@@ -85,6 +104,31 @@
     return *m_onChanged;
 }
 
+void WebExtensionContextProxy::dispatchStorageChangedEvent(const String& onChangedJSON, WebExtensionStorageType storageType, WebExtensionContentWorldType contentWorldType)
+{
+    NSDictionary *onChangedData = parseJSON(onChangedJSON);
+
+    switch (contentWorldType) {
+    case WebExtensionContentWorldType::ContentScript:
+        enumerateContentScriptFramesAndNamespaceObjects([&](WebFrame&, auto& namespaceObject) {
+            namespaceObject.storage().onChanged().invokeListenersWithArgument(onChangedData);
+            namespaceObject.storage().storageAreaForType(storageType).onChanged().invokeListenersWithArgument(onChangedData);
+        });
+        break;
+
+    case WebExtensionContentWorldType::Main:
+        enumerateNamespaceObjects([&](auto& namespaceObject) {
+            namespaceObject.storage().onChanged().invokeListenersWithArgument(onChangedData);
+            namespaceObject.storage().storageAreaForType(storageType).onChanged().invokeListenersWithArgument(onChangedData);
+        });
+        break;
```

Comment:
These could be combined to use `enumerateNamespaceObjects()` and pass the right world for the second optional argument, based on `contentWorldType`. A helper function not get the world based on `contentWorldType` would be good for this too.